src/Project.hs

```diff
@@ -152,26 +169,41 @@ data PullRequest = PullRequest
   , approval            :: Maybe Approval
   , integrationStatus   :: IntegrationStatus
   , integrationAttempts :: [Sha]
+  , integrationChecks   :: Map Sha (Map Check BuildStatus)
```

Comment:
Reading the new code maintaining this map, I was thinking that part of the reason why it's so complex could be the way this type is defined.

Previously, the build status was only relevant in the `Integrated` integration status, and maintained in there.
Now, we have to (somewhat awkwardly) maintain those two states, but where the state in `Integrated` isn't really relevant anymore when there are mandatory checks defined.

Would it be an idea to change the `Integrated` constructor instead to hold this state, rather than having it on the PR level?

Then we could have a single data type holding the progress towards promotion:

```haskell
data OutstandingChecks = AnyCheck BuildStatus | SpecificChecks (Map Check BuildStatus)

data IntegrationStatus = ... | Integrated Sha OutstandingChecks
```

The way I see it, this should make the code a lot more straightforward. For every incoming build status change, in the `AnyCheck` case we'd basically keep doing what we did so far, while in the `SpecificChecks` case we update just the entry corresponding to the context of the build status change.

And where we previously just matched on `Integrated _ BuildSucceeded` for proceeding with a PR, we could now check whether either it's `AnyCheck BuildSucceeded` or `SpecificChecks checks | all (== BuildSucceeded) checks`.

It could also be that I'm missing some inherent complexity in this problem that I haven't taken into account here. Let me know what you think.

Reply:
Good idea. From a birds-eye view, the current complexity comes from trying to keep the `IntegrationStatus` in sync with the checks that have occured at that moment, so essentially trying to fold over all incoming `BuildStatus`es in realtime, while still supporting the case we don't have to fold at all. Your approach moves the fold to places where we pattern match on `Integrated`, which means we have to fold at each pattern match point to get the summarised status.

Intuition tells me that the nicer logic is worth the extra cpu cycles, so I'll try it out!